cleverhans/utils_tf.py

```diff
@@ -109,13 +85,11 @@ def model_train(sess, x, y, predictions, X_train, Y_train, save=False,
     if rng is None:
         rng = np.random.RandomState()
 
-    # Define loss
-    loss = model_loss(y, predictions)
-    if predictions_adv is not None:
-        loss = (loss + model_loss(y, predictions_adv)) / 2
-
+    # Define optimizer
+    loss_value = loss.fprop(x, y, **fprop_args)
     train_step = tf.train.AdamOptimizer(learning_rate=args.learning_rate)
-    train_step = train_step.minimize(loss, var_list=var_list)
+    with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
```

Comment:
Can you comment on what this line does? It might be OK, I just don't know about what this does.
Is there a way that we could avoid using global variables?
A lot of the time for research purposes we want to do things like train two different models in the same process, and global variables / global update lists make it easy to have bugs where we updated both models by accident when we meant to update only one.

Reply:
The globals in tf collections are local to the graph (I verified it by hand just one with a small example)
```
g1 = tf.Graph()
with g1.as_default():
  a1 = tf.get_variable('a', [], tf.float32)
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, tf.assign_add(a1, 1))
  with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
    b1 = 2 * a1
  sess1 = tf.train.MonitoredSession()

g2 = tf.Graph()
with g2.as_default():
  a2 = tf.get_variable('a', [], tf.float32)
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, tf.assign_add(a2, -1))
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, tf.assign_add(a2, +3))
  with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
    b2 = 2 * a2
  sess2 = tf.train.MonitoredSession()

# On graph1
print(sess1.run(a1))
print(sess1.run(b1))  # Increments a1 by 1
print(sess1.run(a1))  # Indeed it was incremented

# Now on graph2
print(sess2.run(a2))
print(sess2.run(b2)) # Increments a2 by 2 (-1 + 3)
print(sess2.run(a2)) # Indeed it was incremented by 2
print(sess1.run(a1)) # However a1 was unchanged since it is in another graph.

# You can also inspect
with g1.as_default(): 
  print(tf.get_collection(tf.GraphKeys.UPDATE_OPS))
  # a single update in the list

with g2.as_default(): 
 print(tf.get_collection(tf.GraphKeys.UPDATE_OPS))
 # two updates in the list (-1 + 3)
```

Reply:
And yes I added a small comment.